src/metrics.hpp

```diff
@@ -286,38 +293,89 @@ class Metrics : public Allocated {
+        }
+        return;
       }
 
-      if (h->total_count == 0) {
-        // There is no data; default to 0 for the stats.
-        snapshot->max = 0;
-        snapshot->min = 0;
-        snapshot->mean = 0;
-        snapshot->stddev = 0;
-        snapshot->median = 0;
-        snapshot->percentile_75th = 0;
-        snapshot->percentile_95th = 0;
-        snapshot->percentile_98th = 0;
-        snapshot->percentile_99th = 0;
-        snapshot->percentile_999th = 0;
-      } else {
-        snapshot->max = hdr_max(h);
-        snapshot->min = hdr_min(h);
-        snapshot->mean = static_cast<int64_t>(hdr_mean(h));
-        snapshot->stddev = static_cast<int64_t>(hdr_stddev(h));
-        snapshot->median = hdr_value_at_percentile(h, 50.0);
-        snapshot->percentile_75th = hdr_value_at_percentile(h, 75.0);
-        snapshot->percentile_95th = hdr_value_at_percentile(h, 95.0);
-        snapshot->percentile_98th = hdr_value_at_percentile(h, 98.0);
-        snapshot->percentile_99th = hdr_value_at_percentile(h, 99.0);
-        snapshot->percentile_999th = hdr_value_at_percentile(h, 99.9);
+      // Refresh interval is in use.  If we've exceeded the interval clear histogram_,
+      // compute a new aggregate histogram and build (and cache) a new snapshot.  Otherwise
+      // just return the cached version.
+      uint64_t now = get_time_since_epoch_ms();
+      if (now - refresh_timestamp_ >= refresh_interval_) {
+
+        hdr_reset(histogram_);
+
+        for (size_t i = 0; i < thread_state_->max_threads(); ++i) {
+          histograms_[i].add(histogram_);
+        }
+
+        cached_snapshot_ = build_new_snapshot(histogram_);
+        refresh_timestamp_ = now;
       }
+
+      copy_snapshot(cached_snapshot_, snapshot);
     }
 
   private:
+
+    void copy_snapshot(Snapshot from, Snapshot* to) const {
+      to->min = from.min;
+      to->max = from.max;
+      to->mean = from.mean;
+      to->stddev = from.stddev;
+      to->median = from.median;
+      to->percentile_75th = from.percentile_75th;
+      to->percentile_95th = from.percentile_95th;
+      to->percentile_98th = from.percentile_98th;
+      to->percentile_99th = from.percentile_99th;
+      to->percentile_999th = from.percentile_999th;
+    }
+
+    Snapshot build_new_snapshot(hdr_histogram* h) const {
+      return Snapshot {
+        hdr_min(h),
+        hdr_max(h),
+        static_cast<int64_t>(hdr_mean(h)),
+        static_cast<int64_t>(hdr_stddev(h)),
+        hdr_value_at_percentile(h, 50.0),
+        hdr_value_at_percentile(h, 75.0),
+        hdr_value_at_percentile(h, 95.0),
+        hdr_value_at_percentile(h, 98.0),
+        hdr_value_at_percentile(h, 99.0),
+        hdr_value_at_percentile(h, 99.9)
+      };
+    }
+
+    // Optimized version of chained build_new_snapshot -> copy_snapshot calls
+    // to avoid creation of an unnecessary intermediate struct
+    void build_and_copy_snapshot(hdr_histogram* h, Snapshot* to) const {
```

Comment:
I'd probably remove the `build` word from this, it's not actually creating a new object like `build_new_snapshot` is.

Maybe `copy_snapshot_to_hdr_hist`?

Reply:
I went with the simpler "histogram_to_snapshot"... hopefully that's clearer.